File: craftbukkit/src/main/java/me/lukiiy/discordBridge/listeners/PlayerEvents.java
package me.lukiiy.discordBridge.listeners;

import me.lukiiy.discordBridge.DiscordBridge;
import me.lukiiy.discordBridge.api.DiscordContext;
import me.lukiiy.discordBridge.api.serialize.DSerial;
import me.lukiiy.discordBridge.event.BridgeMinecraftReceiveEvent;
import me.lukiiy.discordBridge.utils.MemberHelper;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.event.player.PlayerChatEvent;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.event.player.PlayerListener;
import org.bukkit.event.player.PlayerQuitEvent;

public class PlayerEvents extends PlayerListener {
    public void onPlayerChat(PlayerChatEvent e) {
        Player p = e.getPlayer();
        BridgeMinecraftReceiveEvent bridgeEvent = new BridgeMinecraftReceiveEvent(p, e.getMessage());

        Bukkit.getServer().getPluginManager().callEvent(bridgeEvent);
        if (bridgeEvent.isCancelled()) return;

        send(DiscordBridge.getInstance().getConfiguration().getString("format.mc", e.getFormat().replace("%1$s", "(user)").replace("%2$s", "(msg)"))
                .replace("(user)", p.getDisplayName())
                .replace("(msg)", bridgeEvent.getMessage()), true);
    }

    public void onPlayerJoin(PlayerJoinEvent e) {
        send(e.getJoinMessage(), false);
    }

    public void onPlayerQuit(PlayerQuitEvent e) {
        send(e.getQuitMessage(), false);
    }

    private void send(String msg, boolean priority) {
        DiscordBridge instance = DiscordBridge.getInstance();
        DiscordContext context = instance.getContext();

        if (context == null || (priority && !instance.getConfiguration().getBoolean("discord.playerEvents", true)) || msg == null || msg.isEmpty()) return;

        context.sendMessage(DSerial.toDiscord(MemberHelper.fixMentions(msg, context.getGuild())));
    }
}

File: minestom/src/main/kotlin/me/lukiiy/discordBridge/listeners/PlayerEvents.kt
package me.lukiiy.discordBridge.listeners

import me.lukiiy.discordBridge.Main
import me.lukiiy.discordBridge.api.serialize.DSerialAdvnt.toDiscord
import me.lukiiy.discordBridge.utils.MemberHelper.fixMentions
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.serializer.plain.PlainTextComponentSerializer
import net.minestom.server.coordinate.Pos
import net.minestom.server.event.GlobalEventHandler
import net.minestom.server.event.player.AsyncPlayerConfigurationEvent
import net.minestom.server.event.player.PlayerChatEvent
import net.minestom.server.event.player.PlayerDeathEvent
import net.minestom.server.event.player.PlayerDisconnectEvent
import net.minestom.server.instance.Instance

class PlayerEvents(eventHandler: GlobalEventHandler, initialInstance: Instance) {
    val plain = PlainTextComponentSerializer.plainText()

    init {
        eventHandler.apply {
            addListener(PlayerChatEvent::class.java) {
                val p = it.player

                send(Main.config.getOrDefault("messages.discord.format", "")!!
                    .replace("(user)", plain.serialize(p.name))
                    .replace("(msg)", it.rawMessage), true)
            }

            addListener(AsyncPlayerConfigurationEvent::class.java) {
                val p = it.player

                it.spawningInstance = initialInstance
                p.respawnPoint = Pos(0.0, 42.0, 0.0)
                send("${plain.serialize(p.name)} joined", false)
            }

            addListener(PlayerDisconnectEvent::class.java) { send("${plain.serialize(it.player.name)} left", false) }

            addListener(PlayerDeathEvent::class.java) { send("${plain.serialize(it.player.name)} died", false) }
        }
    }

    private fun send(msg: String, priority: Boolean) {
        if ((priority && !Main.config.getBoolean("discord.playerEvents")) || msg.isBlank()) return
        val context = Main.context

        context!!.sendMessage(toDiscord(fixMentions(msg, context.guild)))
    }
}